pkg/data/generic.go

```diff
@@ -79,167 +87,175 @@ func (s *GenericSync) RunContext(ctx context.Context) error {
 	if s.createError != nil {
 		return s.createError
 	}
-	s.loop(ctx.Done())
+
+	store, queue := s.setup(ctx)
+	defer queue.ShutDown()
+
+	go s.loop(ctx, store, queue)
+	<-ctx.Done()
 	return nil
 }
 
-func (s *GenericSync) loop(quit <-chan struct{}) {
+// setup the store and queue for this GenericSync instance
+func (s *GenericSync) setup(ctx context.Context) (cache.Store, workqueue.Interface) {
 
-	defer func() {
-		logrus.Infof("Sync for %v finished. Exiting.", s.ns)
-	}()
-
-	resource := s.client.Resource(schema.GroupVersionResource{
+	baseResource := s.client.Resource(schema.GroupVersionResource{
 		Group:    s.ns.Group,
 		Version:  s.ns.Version,
 		Resource: s.ns.Resource,
 	})
+	var resource dynamic.ResourceInterface = baseResource
+	if s.ns.Namespaced {
+		resource = baseResource.Namespace(metav1.NamespaceAll)
+	}
 
-	delay := backoffMin
+	queue := workqueue.New()
+	store, controller := cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return resource.List(ctx, options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return resource.Watch(ctx, options)
+			},
+		},
+		&unstructured.Unstructured{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if key, err := cache.MetaNamespaceKeyFunc(obj); err == nil {
+					var _ string = key // Tiny sanity check. The code depends on queue keys being strings.
+					queue.Add(key)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if newMeta, err := meta.Accessor(newObj); err == nil {
+					if oldMeta, err := meta.Accessor(oldObj); err == nil {
+						if newMeta.GetResourceVersion() == oldMeta.GetResourceVersion() {
+							return // Avoid sync flood on relist. We don't use resync.
+						}
+					}
+				}
+				if key, err := cache.MetaNamespaceKeyFunc(newObj); err == nil {
+					queue.Add(key)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj); err == nil {
+					queue.Add(key)
+				}
+			},
+		},
+	)
```

Comment:
A lot of `err != nil` are ignore here. Are we OK with this, does it mirror the previous behaviour?

Reply:
The previous version didn't have to deal with this, because it used a typed `Resource.ListWatcher` (https://github.com/open-policy-agent/kube-mgmt/blob/c71de33f3d92de45189561a41a8501dfe6823c52/pkg/data/generic.go#L185) which delivered `unstructured.Unstructured` objects. The `informer` and `store` in the current version deliver `interface{}` instead, which it's uglier and adds a lot of casts and type assertions.

I have trusted all objects to be `unstructured.Unstructured` underneath, since they all come from the same `dynamic.Resource` as before, and dismissed all the errors. But I can log them instead.

Reply:
Do we have log levels here? If so, debug logging would sound fine, what do you think?

Reply:
_But really_, I don't know this code well. If you think it's OK the way it is, let's keep it as it is.

Reply:
A bit of refactoring later, all this has been moved under the type `queueHandler`